### PR TITLE
Created Radix Sort

### PR DESCRIPTION
**Pull request by Rajiv2605**
I have created the code for Radix Sort. I hope that you like it. Please give me suggestions if I could have done anything better as this is my first pull request and would love to learn more.

### CODE REVIEW

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Rename `SIZE` to `size`.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Please don't make all variables global. Instead, declare them in the function where they're required and pass them to other functions when needed.
You should make `size`, `order`, `mult_factor` and `add_factor` local to `main()`.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Also mention the time and space complexity of this function.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Also mention the time and space complexity of this function.

**Review by faheel — CHANGES_REQUESTED** [review]
- Your code has mixed indentation. Please convert each indentation level to 4 spaces. You can configure your text editor to emit 4 spaces whenever you press the tab key.
- Mention the time and space complexity of those functions which are directly related to the algorithm being implemented, which are `count_sort` and `radix_sort` for this program. Click [here](https://github.com/faheel/Algos/blob/master/DataStructures/DisjointSet/DisjointSet.cpp#L28-L42) for an example.
- Don't make all variables global. Instead, declare them in the function where they're required and pass them to other functions when needed.

**Comment by Rajiv2605** [comment]
@faheel I have done all the changes you suggested. Please review it.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: ~`This space is used for count sorting each key and temporarily creating a sorted array.`~
There's no need to mention this under time complexity.

**Review by faheel — COMMENTED** [review]
Looks fine :+1: but I haven't tested it yet. Will do soon.

**Comment by Rajiv2605** [comment]
@faheel I think now the code is fine. Please review it.

**Comment by Rajiv2605** [comment]
@faheel Is the code fine?

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: No need to pass `size` as `values.size()` will give the number of elements (in constant time).
Also, there's no need to pass `order` as `mult_factor` and `add_factor` are being passed.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Change the signature to
```C++
void radix_sort(vector<int>& values, const int mult_factor, const int add_factor, const bool to_show_state)
```

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Simply:
```C++
if (to_show_state)
    display_state(values);
```
`display_state()` is defined in SortingUtils.hpp

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Declaring `output` as
```C++
vector<int> output(values.size());
```
would be simpler and better.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Use a better name for `i`.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: - Change the signature to:
  ```C++
  void count_sort(vector<int>& values, const int I, const int mult_factor, const int add_factor, const bool to_show_state)
  ```
- Use a better name for `I`.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Use a range-based for loop for simplification.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: No need of a loop here. Simply:
```C++
values = output;
```

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: This can be done in `main()` itself. No need for a separate function.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Declare `counter` as:
```C++
array<int, 10> counter{};
```
to be specific to C++.

**Review by faheel — CHANGES_REQUESTED** [review]
Sorry for the late review :sweat_smile:. Here's a summary:
- No need to pass `size` as it can be obtained using `values.size()`
- Use C++ containers such as `std::array` and modern (C++1x) features like [range-based for loops](http://en.cppreference.com/w/cpp/language/range-for), rather than C-style arrays, `malloc()` etc.

File: Sorting/RadixSort.cpp
Comment (ritwik12) [reply Sorting/RadixSort.cpp]: Don't forgot to use` #include<array>` while using `array<int, 10> counter{};`

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: These expressions are hard to understand. Please simplify these.
Also `counter[add_factor + (mult_factor * ((values[i]/I) % 10))]` is used multiple times. I suggest using a function that returns the required reference, and adding comments explaining what the expression and the involved terms are.
Let me know if you need some help implementing the function.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Yes :+1:

**Comment by Rajiv2605** [comment]
@faheel I will do the suggested changes by tomorrow.

File: Sorting/RadixSort.cpp
Comment (Rajiv2605) [reply Sorting/RadixSort.cpp]: Why do I need a function? I could rather use a local variable which would store the value of  `counter[add_factor + (mult_factor * ((values[i]/I) % 10))]` and then I would use that variable everywhere locally.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Using a range-based for loop here will work as intended, so please add it back.
On line 38 however, since you are iterating the container in a reverse order, you cannot use a range-based for loop as it iterates from the beginning until the end.

File: Sorting/RadixSort.cpp
Comment (faheel) [reply Sorting/RadixSort.cpp]: Since the value `add_factor + (mult_factor * ((values[i]/I) % 10))` depends on `values[i]`, or more specifically, just the index `i`, you cannot store it in a variable. You would either need to pass the value `values[i]`, or pass `values` and `i`, along with `add_factor`, `mult_factor` and `extractor` to a function which would return the required value.
As to why use a function for this at all, it is because it would reduce code repetition and make the code shorter and simpler to understand.
**NOTE:** Instead of creating a global function and passing it all the values (`values[i]`, `add_factor`, `mult_factor` and `extractor`), you can use a [lambda expression](http://en.cppreference.com/w/cpp/language/lambda) and pass only `values[i]` as the parameter to further simplify the expression.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Use [`max_element()`](http://en.cppreference.com/w/cpp/algorithm/max_element) from [`algorithm`](http://en.cppreference.com/w/cpp/algorithm) instead of this function.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Rename `m` to `max_value`.

**Review by faheel — CHANGES_REQUESTED** [review]

**Comment by Rajiv2605** [comment]
@faheel I have done the changes.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: @Rajiv2605 You missed this

_Sent from my Samsung GT-I9500 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: @Rajiv2605 I mentioned the following in a previous comment which you may have missed:

Since the value `add_factor + (mult_factor * ((values[i]/I) % 10))` depends on `values[i]`, or more specifically, just the index `i`, you cannot store it in a variable. You would either need to pass the value `values[i]`, or pass `values` and `i`, along with `add_factor`, `mult_factor` and `extractor` to a function which would return the required value.
As to why use a function for this at all, it is because it would reduce code repetition and make the code shorter and simpler to understand.
**NOTE:** Instead of creating a global function and passing it all the values (`values[i]`, `add_factor`, `mult_factor` and `extractor`), you can use a [lambda expression](http://en.cppreference.com/w/cpp/language/lambda) and pass only `values[i]` as the parameter to further simplify the expression.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Can you add a comment here mentioning what `extractor` is?

**Review by faheel — CHANGES_REQUESTED** [review]

File: Sorting/RadixSort.cpp
Comment (Rajiv2605) [reply Sorting/RadixSort.cpp]: @faheel Won't adding a lambda expression would make the code look more complex? Lambda would basically return the expression `add_factor + (mult_factor * ((values[i]/I) % 10))` in-line. Rather, we could just let the expression be there like it is right now. By variable, I meant, inside the loop, we could have something like this:
```
for(int value : values) {
    int x = add_factor + (mult_factor * ((values[i]/extractor) % 10))
    counter[x]++;
}
```

Comment (faheel) [reply Sorting/RadixSort.cpp]: The above would be fine only if the expression was being used once, but it is used 3 times in this function, and so to reduce repetition (following the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)) we should use a function that returns the desired value. But to a function you would need to pass lots of variables to get the desired value. That is why I suggest using a lambda expression as the following:
```C++
int my_lambda = [&] (const int& value) {
    return add_factor + mult_factor * ((value / extractor) % 10);
};
```
Usage example:
```C++
for (const int& value : values)
    counter[my_lambda(value)]++;

...

for (int i = values.size() - 1; i >= 0; i--) {
    output[counter[my_lambda(values[i])] - 1] = values[i];
    ...
}
```
See, it's not at all complex! :smiley:

P. S. Remember to give the lambda a more meaningful name.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: This comment should be removed.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Since the value of `B` is a constant (=10), the time and space complexity of `count_sort()` and the space complexity of `radix_sort()` would be,
**_O_**(N + 10)
= **_O_**(N), where N is the number of keys (values)

**Review by faheel — CHANGES_REQUESTED** [review]

**Comment by Rajiv2605** [comment]
@faheel I have done all the changes:+1: I guess. My compiler was throwing a lot of errors when I had declared the variable equated to the lambda expression as `int`. But, when I changed it to `auto` type, it worked! What would have caused the problem?

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Have this local to `count_sort()` so that you wouldn't need to pass `mult_factor`, `add_factor` and `extractor` as arguments, they will automatically be captured as references if you declare the lambda as in the example I gave.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: @Rajiv2605 You missed this.

**Review by faheel — CHANGES_REQUESTED** [review]

File: Sorting/RadixSort.cpp
Comment (faheel) [reply Sorting/RadixSort.cpp]: @Rajiv2605 You missed this.

**Comment by faheel** [comment]
About the lambda's type, I don't know why you were getting errors on `int` as there's nothing wrong with it. Did you try with `-std=c++14` flag?

**Comment by faheel** [comment]
There have been lots of suggestions recently by interested contributors to add more algorithms and data structures to the project, which I have currently put on hold because I want to close issue #74 first, as it would require a complete reorganisation of files and some refactoring of code.

I previously thought that I would work on #74 once the currently active PRs are merged, but I feel that it would take much longer that way. So, if it's not too inconvenient for you, I would like to work on issue #74 as soon as possible. The remaining work on this PR (and others) can be carried on afterwards. Once the project is refactored, all you would need to do is `git fetch upstream` and move the code that you wrote to the new appropriate directory.

So, what do you say?

File: Sorting/RadixSort.cpp
Comment (Rajiv2605) [reply Sorting/RadixSort.cpp]: I have done it in the next line.

**Comment by Rajiv2605** [comment]
:+1: I'll wait for you to finish your PR. Let me know when you finish it.

File: Sorting/RadixSort.cpp
Comment (faheel) [reply Sorting/RadixSort.cpp]: All the explanation about the complexity is not needed here. Simply write
```C++
Time complexity
---------------
O(N), where N is the number of keys.

Space complexity
----------------
O(N), where N is the number of keys.
```
Also elsewhere.

File: Sorting/RadixSort.cpp
Comment (faheel) [reply Sorting/RadixSort.cpp]: You probably missed this.
Click on the `Files changed` tab for previous details.

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: This should now be `#include "sorting/utils.hpp"`

File: Sorting/RadixSort.cpp
Comment (faheel) [inline Sorting/RadixSort.cpp]: Take `values` by reference to avoid unnecessary copies of it being made.

**Review by faheel — CHANGES_REQUESTED** [review]
- The file should now be placed in `C++/source/sorting/` with the name `radix_sort.cpp`.
- Take all `const` arguments as references to avoid unnecessary copies of them being made.

File: C++/source/sorting/radix_sort.cpp
Comment (faheel) [inline C++/source/sorting/radix_sort.cpp]: All of these parameters should taken as references too.

File: C++/source/sorting/radix_sort.cpp
Comment (faheel) [inline C++/source/sorting/radix_sort.cpp]: Capture list should be `[&]`

**Review by faheel — CHANGES_REQUESTED** [review]

**Comment by Rajiv2605** [comment]
@faheel Is it fine now?

**Comment by faheel** [comment]
Oh no, it doesn't seem to work :disappointed:. With the input `3 6 2 0 -5 4 -8 -6 4 1 7 5 8 6 7 5` (size 16) it gives a seg fault when sorting in increasing order and gives the wrong values when sorting in decreasing order.

**Comment by Rajiv2605** [comment]
@faheel Oh... :confused: It worked with many inputs I tried... I'll check... :disappointed:

**Comment by Rajiv2605** [comment]
@faheel The code is unable to locate `sorting/utils.cpp`. Also, the reason you got error is that Radix Sort's general algorithm, i.e., the one I have implemented doesn't sort negative numbers. The code I had pushed which used `SortingUtils.hpp` works perfectly for non-negative numbers.

**Comment by faheel** [comment]
Ah, right, it will only sort positive values. My bad! Please mention this fact in the description comment.

Regarding "not being able to locate `sorting/utils.hpp`", use the command **`make all`** from the `C++` folder to compile the sources. It's much more convenient and gives no issues. More details can be found in the [Compiling](../tree/master/C%2B%2B#compiling) section of the README.

**Comment by Rajiv2605** [comment]
@faheel I have added it to the comments:v:

**Comment by faheel** [comment]
@Rajiv2605 Thanks a lot for your time and contribution! :smiley:

**Comment by Rajiv2605** [comment]
Thanks!

**Pull request merged**
